File: 2665-minimum-time-to-repair-cars/2665-minimum-time-to-repair-cars.py
class Solution:
    def isPossible(self, ranks, cars, min_time):
        cnt = 0
        for r in ranks:
            cnt += int(math.sqrt(min_time / r))  # Number of cars mechanic can repair
            if cnt >= cars:
                return True
        return False
        
    def repairCars(self, ranks: List[int], cars: int) -> int:
        n = len(ranks)
        c = math.ceil(cars / n) ** 2

        l = min(ranks) * c  # Minimum possible time
        r = max(ranks) * c  # Maximum possible time

        while l <= r:
            mid = (l + r) // 2
            if self.isPossible(ranks, cars, mid):
                r = mid - 1  # Try to minimize time
            else:
                l = mid + 1  # Increase time
        return l

File: 2665-minimum-time-to-repair-cars/2665-minimum-time-to-repair-cars.java
class Solution {
    public boolean isPossible(int[] ranks,int cars,long min){
        int cnt = 0;
        for(int i:ranks){
            cnt += (int) Math.sqrt(min / i);
            if(cnt >= cars)
                return true;
        }
        return false;
    }
    public long repairCars(int[] ranks, int cars) {
        int n = ranks.length;
        long l = Integer.MAX_VALUE,r = 0,c = (long) Math.pow(Math.ceil((double)cars/n),2);
        for(int i:ranks){
            l = Math.min(l,i);
            r = Math.max(r,i);
        }
        l = l * c;
        r = r * c;
        while(l <= r){
            long mid = (l + r) / 2;
            if(isPossible(ranks,cars,mid)){
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return l;
    }
}